Foundation/URLCache.swift

```diff
@@ -225,36 +503,91 @@ open class URLCache : NSObject {
         stored with the given request.
         @param request the NSURLRequest to use as a key for the lookup.
     */
-    open func removeCachedResponse(for request: URLRequest) { NSUnimplemented() }
+    open func removeCachedResponse(for request: URLRequest) {
+        guard let identifier = identifier(for: request) else { return }
+        
+        inMemoryCacheLock.performLocked {
+            if inMemoryCacheContents[identifier] != nil {
+                inMemoryCacheOrder.removeAll(where: { $0 == identifier })
+                inMemoryCacheContents.removeValue(forKey: identifier)
+            }
+        }
+        
+        if let oldURL = diskContentsURL(for: request) {
+            try? FileManager.default.removeItem(at: oldURL)
+        }
+    }
     
     /*! 
         @method removeAllCachedResponses
         @abstract Clears the given cache, removing all NSCachedURLResponse
         objects that it stores.
     */
-    open func removeAllCachedResponses() { NSUnimplemented() }
+    open func removeAllCachedResponses() {
+        inMemoryCacheLock.performLocked {
+            inMemoryCacheContents = [:]
+            inMemoryCacheOrder = []
+        }
+        
+        evictFromDiskCache(maximumSize: 0)
+    }
     
     /*!
      @method removeCachedResponsesSince:
      @abstract Clears the given cache of any cached responses since the provided date.
      */
-    open func removeCachedResponses(since date: Date) { NSUnimplemented() }
+    open func removeCachedResponses(since date: Date) {
+        inMemoryCacheLock.performLocked { // Memory cache:
+            var identifiersToRemove: Set<String> = []
+            for entry in inMemoryCacheContents {
+                if entry.value.date > date {
+                    identifiersToRemove.insert(entry.key)
+                }
+            }
+            
+            for toRemove in identifiersToRemove {
+                inMemoryCacheContents.removeValue(forKey: toRemove)
+            }
+            inMemoryCacheOrder.removeAll { identifiersToRemove.contains($0) }
+        }
+        
+        do { // Disk cache:
+            var urlsToRemove: [URL] = []
+            enumerateDiskEntries { (entry, stop) in
+                if entry.date > date {
+                    urlsToRemove.append(entry.url)
+                }
+            }
+            
+            for url in urlsToRemove {
+                try? FileManager.default.removeItem(at: url)
+            }
```

Comment:
From the implementation of `enumerateDiskEntries`, since the `contentsOfDirectory(at:…)` are generated before invoking the block, I think you can merge both loops into one and forget about `urlsToRemove`.

```swift
enumerateDiskEntries { entry, _ in
  if entry.date > data {
    try? FileManager.default.removeItem(at: url)
  }
}
```